pkg/pool-weighted/contracts/managed/ManagedPoolTokenLib.sol

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/solidity-utils/openzeppelin/IERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ERC20.sol";
+
+import "../lib/GradualValueChange.sol";
+import "../lib/WeightCompression.sol";
+
+library ManagedPoolTokenLib {
+    using WordCodec for bytes32;
+    using FixedPoint for uint256;
+    using WeightCompression for uint256;
+
+    // Denormalized weights are stored using the WeightCompression library as a percentage of the maximum absolute
+    // denormalized weight: independent of the current _denormWeightSum, which avoids having to recompute the denorm
+    // weights as the sum changes.
+    uint256 private constant _MAX_DENORM_WEIGHT = 1e22; // FP 10,000
+
+    // Store scaling factor and start/end denormalized weights for each token
+    // Mapping should be more efficient than trying to compress it further
+    // [ 123 bits |  5 bits  |  64 bits   |   64 bits    |
+    // [ unused   | decimals | end denorm | start denorm |
+    // |MSB                                           LSB|
+    uint256 private constant _START_DENORM_WEIGHT_OFFSET = 0;
+    uint256 private constant _END_DENORM_WEIGHT_OFFSET = 64;
+    uint256 private constant _DECIMAL_DIFF_OFFSET = 128;
+
+    uint256 private constant _DENORM_WEIGHT_WIDTH = 64;
+    uint256 private constant _DECIMAL_DIFF_WIDTH = 5;
+
+    // Getters
+
+    /**
+     * @notice Returns the token's scaling factor.
+     */
+    function getTokenScalingFactor(bytes32 tokenState) internal pure returns (uint256) {
+        uint256 decimalsDifference = tokenState.decodeUint(_DECIMAL_DIFF_OFFSET, _DECIMAL_DIFF_WIDTH);
+
+        // This is equivalent to `10**(18+decimalsDifference)` but this form optimizes for 18 decimal tokens.
+        return FixedPoint.ONE * 10**decimalsDifference;
```

Comment:
It's an EXP opcode which has a cost of `10 + 50 * exponent_byte_size`. I was thinking that we could do the multiplication for free if we shove `FixedPoint.ONE` up in the exponent but then we'd always have a non-zero exponent and pay 50 rather than 10 gas for most tokens.